test/smart-contract-client.test.js

```diff
@@ -0,0 +1,101 @@
+import { assertEquals, assertRejects, assert } from 'zinnia:assert'
+import { getMinerPeerIdFromSmartContract } from '../lib/smart-contract-client.js'
+import { getMinerPeerId } from '../lib/miner-info.js'
+import { test } from 'zinnia:test'
+
+const mockPeerIdResponse = {
+  peerID: '12D3KooWGQmdpbssrYHWFTwwbKmKL3i54EJC9j7RRNb47U9jUv1U',
+  signature: '0x1234567890abcdef'
+}
+
+const mockEmptyPeerIdResponse = {
+  peerID: '',
+  signature: '0x'
+}
+
+// Mock contract factory
+function createMockContract (mockResponses) {
+  return {
+    getPeerData: async (minerId) => {
+      const response = mockResponses[minerId]
+      if (!response) {
+        throw new Error(`Miner ID ${minerId} not found in contract`)
+      }
+      return response
+    }
+  }
+}
+
+test('getMinerPeerIdFromSmartContract returns peer ID for valid miner ID', async () => {
+  // Create mock contract with predefined responses
+  const peerId = 12345
+  const mockContract = createMockContract({
+    [peerId]: mockPeerIdResponse
+  })
+
+  const actualPeerId = await getMinerPeerIdFromSmartContract(`f0${peerId}`, {
+    smartContract: mockContract
+  })
+
+  assertEquals(actualPeerId, mockPeerIdResponse.peerID)
+})
+
+test('getMinerPeerIdFromSmartContract returns correct peer id for miner f03303347', async () => {
+  const peerId = await getMinerPeerIdFromSmartContract('f03303347')
+  assertEquals(typeof peerId, 'string', 'Expected peerId to be a string')
+  assertEquals(peerId, '12D3KooWJ91c6xQshrNe7QAXPFAaeRrHWq2UrgXGPf8UmMZMwyZ5')
+})
+
+test('getMinerPeerIdFromSmartContract returns empty string for miner ID with no peer ID', async () => {
+  // Create mock contract with predefined responses
+  const peerId = 99999
+  const mockContract = createMockContract({
+    [peerId]: mockEmptyPeerIdResponse
+  })
+
+  const actualPeerId = await getMinerPeerIdFromSmartContract(`f0${peerId}`, {
+    smartContract: mockContract
+  })
+
+  assertEquals(actualPeerId, '')
+})
+
+test('getMinerPeerIdFromSmartContract throws error for non-existent miner ID', async () => {
```

Comment:
This does not make sense to me.

The test above says that the function returns an empty string when there is no peer id set.

Under which conditions would the smart contract invocation fail?

When I look at the smart contract source, I don't see any assertions that could cause an error:

https://github.com/filecoin-project/curio/blob/395bc47d0f585cbc869fd4671dc05b1b2f4b18c2/market/ipni/spark/sol/contracts/MinerPeerIDMapping.sol#L81-L95

What am I missing?

Reply:
> Let's make sure we have an integration-level test for the case "miner does not have any PeerID registered in the smart contract" - this test should call the real Glif RPC API.
> 
> We want to use a miner ID that will never be an actual miner ID running Curio. For example, we can use one of the known client IDs as the miner ID - I think it's extremely unlikely that anybody would share the same identity both for a client and a storage provider. You can find client IDs in Spark data (recall the work you did on per-client RSR) or on websites like https://filecoin.tools

@NikolasHaimerl PTAL ☝🏻